src/OpenLoco/GameCommands/ChangeCompanyColour.cpp

```diff
@@ -0,0 +1,76 @@
+#include "../Audio/Audio.h"
+#include "../CompanyManager.h"
+#include "../GameException.hpp"
+#include "../OpenLoco.h"
+#include "../Ui/WindowManager.h"
+#include "../Ui/WindowType.h"
+#include "GameCommands.h"
+
+using namespace OpenLoco::Interop;
+
+namespace OpenLoco::GameCommands
+{
+    // 0x0043483D
+    static uint32_t changeCompanyColour(uint8_t flags, uint8_t type, bool toggleMode, CompanyId_t companyId, bool isSecondary, uint8_t value)
+    {
+        GameCommands::setExpenditureType(ExpenditureType::Miscellaneous);
+        GameCommands::setPosition({ static_cast<int16_t>(0x8000), 0, 0 });
+
+        auto* company = CompanyManager::get(companyId);
+
+        if ((flags & Flags::apply) == 1)
+        {
+            // Toggling vehicle palette
+            if (toggleMode)
+            {
+                if (value)
+                    company->customVehicleColoursSet |= (1 << type);
+                else
+                    company->customVehicleColoursSet &= ~(1 << type);
+            }
+            // Setting a colour
+            else
+            {
+                ColourScheme* colours;
+                if (type == 0)
+                    colours = &company->mainColours;
+                else
+                    colours = &company->vehicleColours[type - 1];
+
+                if (!isSecondary)
+                    colours->primary = value;
+                else
+                    colours->secondary = value;
+            }
+
+            company->updateVehicleColours();
+            CompanyManager::updateColours();
+            company->updateHeadquartersColour();
+            Ui::WindowManager::invalidate(Ui::WindowType::company);
+            return 0;
+        }
+        else
+        {
+            if (!sub_431E6A(companyId, nullptr))
+                return GameCommands::FAILURE;
+
+            if (toggleMode || type > 0 || isSecondary)
+                return 0;
+
+            // Check whether the requested colour is available
+            uint32_t availableColours = 0x7FFFFFFFF & ~CompanyManager::competingColourMask(companyId);
```

Comment:
?? why `0x7FFFFFFFF` the value is 32bit? Why are you inverting the value? 

Surely all you need to do is get the mask and test for if valid:

```
auto unavailableColours = CompanyManager::competingColourMask(companyId);
if (unavailableColours & (1 << value)){
    setErrorText(StringIds::empty);
    return GameCommands::FAILURE;
}
```

Reply:
I was copying what vanilla was doing, but your solution looks nicer, yeah.